File: backend/app/schemas/approval.py
"""
簽核系統 Pydantic Schemas
"""
from __future__ import annotations
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel


# ── 關卡 ─────────────────────────────────────────────────────────────────────

class StepOut(BaseModel):
    id: str
    step_order: int
    approver_id: str
    approver_name: str
    approver_email: str
    status: str                    # pending | approved | rejected
    decided_at: Optional[datetime]
    comment: str

    class Config:
        from_attributes = True


# ── 歷程 ─────────────────────────────────────────────────────────────────────

class ActionOut(BaseModel):
    id: str
    step_id: Optional[str]
    actor: str
    actor_id: str
    action: str
    note: str
    created_at: datetime

    class Config:
        from_attributes = True


# ── 附件 ─────────────────────────────────────────────────────────────────────

class FileOut(BaseModel):
    id: str
    orig_name: str
    content_type: str
    size_bytes: int
    uploaded_by: str
    uploaded_at: datetime

    class Config:
        from_attributes = True


# ── 主單清單 ─────────────────────────────────────────────────────────────────

class ApprovalListItem(BaseModel):
    id: str
    subject: str
    requester: str
    requester_id: str
    status: str
    current_step: int
    submitted_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ApprovalListResponse(BaseModel):
    items: List[ApprovalListItem]
    total: int


# ── 主單詳情 ─────────────────────────────────────────────────────────────────

class ApprovalDetail(BaseModel):
    id: str
    subject: str
    description: str
    confidential: str              # 僅 admin/申請人/簽核人可見
    requester: str
    requester_id: str
    requester_dept: str
    status: str
    current_step: int
    view_scope: str
    publish_memo: int
    submitted_at: datetime
    updated_at: datetime
    steps: List[StepOut]
    actions: List[ActionOut]
    attachments: List[FileOut]
    can_act: bool                  # 是否可簽核（目前輪到我）
    can_manage: bool               # 是否可管理串簽（申請人或 admin）

    class Config:
        from_attributes = True


# ── 新增簽核單 ────────────────────────────────────────────────────────────────

class ApproverIn(BaseModel):
    user_id: str                   # Portal user.id（可為空字串代表純姓名輸入）
    name: str                      # full_name
    email: str = ""

class ApprovalCreate(BaseModel):
    subject: str
    description: str = ""
    confidential: str = ""
    requester_dept: str = ""
    view_scope: str = "restricted"
    publish_memo: int = 0
    approver_chain: List[ApproverIn]


# ── 簽核動作 ─────────────────────────────────────────────────────────────────

class ApprovalActionRequest(BaseModel):
    action: str    # approve | reject
    comment: str


# ── 調整關卡順序 ─────────────────────────────────────────────────────────────

class ReorderRequest(BaseModel):
    order: List[str]               # step id 清單，新順序


# ── 新增關卡（已流轉中插入） ──────────────────────────────────────────────────

class AddStepRequest(BaseModel):
    approver: ApproverIn
    insert_after: int = -1         # 插入在哪一關之後（-1=加在最後）


# ── 搜尋結果 ─────────────────────────────────────────────────────────────────

class ApprovalSearchItem(BaseModel):
    id: str
    subject: str
    requester: str
    status: str
    current_step: int
    submitted_at: datetime
    preview: str                         # description 前 120 字（純文字）
    current_approver_name: str = ""      # 目前待簽關卡的簽核人姓名（pending 時才有值）

File: frontend/src/types/approval.ts
/**
 * 簽核系統 TypeScript 型別定義
 * 對應後端 /api/v1/approvals
 */

export type ApprovalStatus = 'pending' | 'approved' | 'rejected'
export type ViewScope     = 'org' | 'restricted' | 'top_secret'
export type StepStatus    = 'pending' | 'approved' | 'rejected'

// ── 關卡 ─────────────────────────────────────────────────────────────────────

export interface ApprovalStep {
  id: string
  step_order: number
  approver_id: string
  approver_name: string
  approver_email: string
  status: StepStatus
  decided_at: string | null
  comment: string
}

// ── 歷程 ─────────────────────────────────────────────────────────────────────

export interface ApprovalAction {
  id: string
  step_id: string | null
  actor: string
  actor_id: string
  action: string
  note: string
  created_at: string
}

// ── 附件 ─────────────────────────────────────────────────────────────────────

export interface ApprovalFile {
  id: string
  orig_name: string
  content_type: string
  size_bytes: number
  uploaded_by: string
  uploaded_at: string
}

// ── 清單項目 ─────────────────────────────────────────────────────────────────

export interface ApprovalListItem {
  id: string
  subject: string
  requester: string
  requester_id: string
  status: ApprovalStatus
  current_step: number
  submitted_at: string
  updated_at: string
}

export interface ApprovalListResponse {
  items: ApprovalListItem[]
  total: number
}

// ── 詳情 ─────────────────────────────────────────────────────────────────────

export interface ApprovalDetail {
  id: string
  subject: string
  description: string
  confidential: string
  requester: string
  requester_id: string
  requester_dept: string
  status: ApprovalStatus
  current_step: number
  view_scope: ViewScope
  publish_memo: number
  submitted_at: string
  updated_at: string
  steps: ApprovalStep[]
  actions: ApprovalAction[]
  attachments: ApprovalFile[]
  can_act: boolean
  can_manage: boolean
}

// ── 搜尋結果 ─────────────────────────────────────────────────────────────────

export interface ApprovalSearchItem {
  id: string
  subject: string
  requester: string
  status: ApprovalStatus
  current_step: number
  submitted_at: string
  preview: string
  current_approver_name: string   // 目前待簽關卡的簽核人姓名（pending 時有值）
}

// ── 可選簽核人 ────────────────────────────────────────────────────────────────

export interface ApproverOption {
  user_id: string
  name: string
  email: string
}

// ── 建立表單用 ────────────────────────────────────────────────────────────────

export interface ApproverIn {
  user_id: string
  name: string
  email: string
}

export interface ApprovalCreatePayload {
  subject: string
  description: string
  confidential: string
  requester_dept: string
  view_scope: ViewScope
  publish_memo: number
  approver_chain: ApproverIn[]
}

// ── 查詢篩選 ─────────────────────────────────────────────────────────────────

export interface ApprovalFilters {
  scope?: 'all' | 'mine' | 'todo'
  status?: 'all' | 'pending' | 'approved' | 'rejected'
  q?: string
  date_from?: string
  date_to?: string
  page?: number
  per_page?: number
}
